File: backend_python/app/services/auth_service.py
"""
Authentication service
"""
from typing import Optional
from app.models.user import User, Role, UserRole
from app.utils.password import hash_password, verify_password
from app.utils.jwt import sign_token
from app.utils.refresh_token import generate_refresh_token, create_refresh_token
from app.middleware.error_handler import HttpError
from app.config.env import settings


class AuthService:
    """Authentication service"""
    
    @staticmethod
    async def register(
        email: str,
        password: str,
        full_name: str,
        phone: Optional[str] = None,
        role: str = "USER",
        device_id: Optional[str] = None,
        ip_address: Optional[str] = None,
        user_agent: Optional[str] = None,
    ) -> dict:
        """Register a new user"""
        # Check if user exists
        existing_user = await User.find_one({"email": email})
        if existing_user:
            raise HttpError(409, "Email already registered")
        
        # Hash password
        password_hash = hash_password(password)
        user_role = UserRole(role) if role in ["USER", "PROVIDER"] else UserRole.USER
        
        # Create user
        user = User(
            email=email,
            passwordHash=password_hash,
            fullName=full_name,
            phone=phone,
        )
        await user.insert()
        
        # Create role
        role_obj = Role(role=user_role, userId=str(user.id))
        await role_obj.insert()
        
        # Generate refresh token
        refresh_token = generate_refresh_token()
        await create_refresh_token(
            user_id=str(user.id),
            token=refresh_token,
            device_id=device_id,
            ip_address=ip_address,
            user_agent=user_agent,
        )
        
        return await AuthService._build_auth_response(user, refresh_token)
    
    @staticmethod
    async def login(
        email: str,
        password: str,
        device_id: Optional[str] = None,
        ip_address: Optional[str] = None,
        user_agent: Optional[str] = None,
    ) -> dict:
        """Login user"""
        # Find user
        user = await User.find_one({"email": email})
        if not user:
            raise HttpError(401, "Invalid credentials")
        
        # Load roles
        roles = await Role.find({"userId": str(user.id)}).to_list()
        
        # Verify password
        if not verify_password(user.password_hash, password):
            raise HttpError(401, "Invalid credentials")
        
        # Generate refresh token
        refresh_token = generate_refresh_token()
        await create_refresh_token(
            user_id=str(user.id),
            token=refresh_token,
            device_id=device_id,
            ip_address=ip_address,
            user_agent=user_agent,
        )
        
        return await AuthService._build_auth_response(user, refresh_token)
    
    @staticmethod
    async def get_profile(user_id: str) -> dict:
        """Get user profile"""
        user = await User.get(user_id)
        if not user:
            raise HttpError(404, "User not found")
        
        await user.fetch_all_links()
        roles = await Role.find({"userId": str(user.id)}).to_list()
        
        return {
            "id": str(user.id),
            "email": user.email,
            "fullName": user.full_name,
            "phone": user.phone,
            "roles": [{"role": role.role.value} for role in roles],
            "createdAt": user.created_at.isoformat(),
        }
    
    @staticmethod
    async def _build_auth_response(user: User, refresh_token: str) -> dict:
        """Build authentication response"""
        # Get user roles
        roles = await Role.find({"userId": str(user.id)}).to_list()
        role_values = [role.role.value for role in roles]
        
        payload = {
            "id": str(user.id),
            "email": user.email,
            "roles": role_values,
        }
        
        access_token = sign_token(payload)
        
        return {
            "accessToken": access_token,
            "refreshToken": refresh_token,
            "expiresIn": settings.jwt_expires_in,
            "refreshTokenExpiresIn": "7d",
            "user": {
                "id": str(user.id),
                "email": user.email,
                "fullName": user.full_name,
                "phone": user.phone,
                "roles": [{"role": role} for role in role_values],
                "createdAt": user.created_at.isoformat(),
            },
        }

File: backend/src/modules/auth/auth.service.ts
import type { PrismaClient, User } from '@prisma/client';

import { env } from '../../config/env.js';
import { HttpError } from '../../middleware/errorHandler.js';
import {
  createRefreshToken,
  generateRefreshToken,
} from './refreshToken.service.js';
import { signToken } from '../../utils/jwt.js';
import { hashPassword, verifyPassword } from '../../utils/password.js';

export async function register(
  prisma: PrismaClient,
  data: {
    email: string;
    password: string;
    fullName: string;
    phone?: string;
    role?: 'USER' | 'PROVIDER';
    deviceId?: string;
    ipAddress?: string;
    userAgent?: string;
  }
) {
  const exists = await prisma.user.findUnique({ where: { email: data.email } });
  if (exists) {
    throw new HttpError(409, 'Email already registered');
  }

  const passwordHash = await hashPassword(data.password);
  const role = data.role || 'USER';

  const user = await prisma.user.create({
    data: {
      email: data.email,
      passwordHash,
      fullName: data.fullName,
      phone: data.phone,
      roles: {
        create: [{ role: role as 'USER' | 'PROVIDER' }],
      },
    },
    include: { roles: true },
  });

  // Generate refresh token
  const refreshToken = generateRefreshToken();
  await createRefreshToken(prisma, user.id, refreshToken, {
    deviceId: data.deviceId,
    ipAddress: data.ipAddress,
    userAgent: data.userAgent,
  });

  return buildAuthResponse(user, refreshToken);
}

export async function login(
  prisma: PrismaClient,
  data: {
    email: string;
    password: string;
    deviceId?: string;
    ipAddress?: string;
    userAgent?: string;
  }
) {
  const user = await prisma.user.findUnique({
    where: { email: data.email },
    include: { roles: true },
  });

  if (!user) {
    throw new HttpError(401, 'Invalid credentials');
  }

  const valid = await verifyPassword(user.passwordHash, data.password);
  if (!valid) {
    throw new HttpError(401, 'Invalid credentials');
  }

  // Generate refresh token
  const refreshToken = generateRefreshToken();
  await createRefreshToken(prisma, user.id, refreshToken, {
    deviceId: data.deviceId,
    ipAddress: data.ipAddress,
    userAgent: data.userAgent,
  });

  return buildAuthResponse(user, refreshToken);
}

function buildAuthResponse(
  user: User & { roles: { role: string }[] },
  refreshToken: string
) {
  const payload = {
    id: user.id,
    email: user.email,
    roles: user.roles.map((role) => role.role),
  };

  const accessToken = signToken(payload);

  return {
    accessToken,
    refreshToken,
    expiresIn: env.jwtExpiresIn,
    refreshTokenExpiresIn: '7d', // 7 days
    user: {
      id: user.id,
      email: user.email,
      fullName: user.fullName,
      phone: user.phone,
      roles: user.roles.map((role) => ({ role: role.role })),
      createdAt: user.createdAt,
    },
  };
}

